### PR TITLE
Use file-based service discovery for prometheus 

### PR DESCRIPTION
**Pull request by ZhengzheYang**
This PR aims to use file_sd_config for service discovery in Prometheus. This could allow workloads entries that do not have endpoints in k8s could be picked up by Prometheus, such as VMs.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[X] Policies and Telemetry
[ ] Security
[X] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by ZhengzheYang** [comment]
@douglas-reid @howardjohn Could you please take a look? This is the initial attempt to mount the endpoints to prometheus. I now update the configmap in deployment of the pod, but we could decide where the update logic should be (either when workload is being created or add-to-mesh istioctl, or somewhere elese). Could use some early feedback to see if I am on the right track. The Prometheus could now query and report with `reporter=destination`, which is not the case before.

File: tests/integration/telemetry/stats/prometheus/http/stats.go [hunk 1]
```diff
@@ -170,6 +214,10 @@ func buildQuery() (sourceQuery, destinationQuery, appQuery string) {
 		"source_workload_namespace":      ns.Name(),
 	}
 	for k, v := range labels {
+		// These two fields remain unknown so we skip them for VMs
+		if destination == "vm" && (k == "destination_app" || k == "destination_version") {
+			continue
+		}
```
Comment (ZhengzheYang) [inline tests/integration/telemetry/stats/prometheus/http/stats.go:220]: I tried to relabel the labels but these two fields are still unknown. Is there anyway to let the prometheus know the app and version info? Maybe by relabeling?

**Comment by howardjohn** [comment]
I am a bit confused by the goals of this as it seems we are implementing this in the tests only? But then we create an empty config map in the real code base. is this meant for testing only?

**Comment by douglas-reid** [comment]
> I am a bit confused by the goals of this as it seems we are implementing this in the tests only? But then we create an empty config map in the real code base. is this meant for testing only?

I think this is just proof-of-concept at the moment, based on the request for feedback.

It seems like we could have the bit that is creating WEs update the config map as part of registration. I don't know if that is the ideal place, but it could work.

File: pkg/test/framework/components/echo/kube/instance.go [hunk 2]
```diff
@@ -313,6 +319,35 @@ func workloadHasSidecar(cfg echo.Config, podName string) bool {
...
+		return err
+	}
+	if currentCM.Data == nil {
+		currentCM.Data = make(map[string]string)
+	}
+
+	staticConfig := `
+- targets:
+  - %s
+  labels:
```
Comment (douglas-reid) [inline pkg/test/framework/components/echo/kube/instance.go:339]: i wouldn't worry about doing any labeling. Rely on that information to come from the proxies themselves. We don't need to match the k8s discovery stuff (in fact, we've tried to remove the extra stuff that adds in previous iterations, etc.)

File: tests/integration/telemetry/stats/prometheus/http/stats.go (hunk 1 above)
Comment (douglas-reid) [reply tests/integration/telemetry/stats/prometheus/http/stats.go:220]: app and version should could from the metrics on the proxies themselves. we shouldn't try to relabel. for tests, we can simulate this by adding:

```json
"LABELS": {
  "app": "server",
  "version": "v1"
}
```

to the custom bootstrap information.

Comment (douglas-reid) [reply tests/integration/telemetry/stats/prometheus/http/stats.go:220]: or by setting the `ISTIO_METAJSON_LABELS` env var.

**Comment by ZhengzheYang** [comment]
> I am a bit confused by the goals of this as it seems we are implementing this in the tests only? But then we create an empty config map in the real code base. is this meant for testing only?

As @douglas-reid said, this is just to prove it works. We could decide where we want to update the configmap and I will update and request another review.

File: pkg/test/framework/components/echo/kube/instance.go (hunk 2 above)
Comment (ZhengzheYang) [reply pkg/test/framework/components/echo/kube/instance.go:339]: Understood!

File: tests/integration/telemetry/stats/prometheus/http/stats.go (hunk 1 above)
Comment (ZhengzheYang) [reply tests/integration/telemetry/stats/prometheus/http/stats.go:220]: Thanks for the suggestion! But what I am confused of is, why we have to simulate for tests? Won't test query info from envoy as well? From port 15090?

**Comment by ZhengzheYang** [comment]
@howardjohn @douglas-reid I updated the logic to the workload instance handler in the kube controller. Let me know if there's a better place for it.

**Comment by ZhengzheYang** [comment]
Not passing the multicluster test due to concurrent map writing. Fixing it.

File: tests/integration/telemetry/stats/prometheus/http/stats.go (hunk 1 above)
Comment (douglas-reid) [reply tests/integration/telemetry/stats/prometheus/http/stats.go]: envoy gets the info on pod labels from its bootstrap config. typically, they are provided by sidecar injector. for vms, we have to hand supply them (for the moment).

Comment (ZhengzheYang) [reply tests/integration/telemetry/stats/prometheus/http/stats.go]: Thanks! That makes sense.

**Comment by douglas-reid** [comment]
I'm excited to see this working!  Thanks for tackling this feature.

I think the next thing would be to draft a quick RFC on the support, referencing this PR as a working approach. It looks like we have some test coverage, which I think can be part of the required test plan. I think we can then iterate quickly on how we want to document this for users, etc.

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 3]
```diff
@@ -818,6 +821,34 @@ func (c *Controller) WorkloadInstanceHandler(si *model.WorkloadInstance, event m
 	default: // add or update
 		c.workloadInstancesByIP[si.Endpoint.Address] = si
 	}
+
+	// Update the config map for workload's service discovery as well.
+	// External workloads have no endpoint registered in k8s, so the traffic sent to these non-k8s workloads
+	// won't be picked up by Prometheus. We update the workload entries' endpoints to a config map so that
+	// Prometheus could mount the map and read the endpoints.
+	wleName := createNameFromAddr(si.Endpoint.Address)
+	fileSDConfig := c.GetOrCreatePromSDConfigMap()
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:830]: fileSDConfig  can be nil here

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 4]
```diff
@@ -818,6 +821,34 @@ func (c *Controller) WorkloadInstanceHandler(si *model.WorkloadInstance, event m
...
+			fileSDConfig.Data = make(map[string]string)
+		}
+		staticConfig := `
+- targets:
+  - %s
+`
+		fileSDConfig.Data[fmt.Sprintf("%s.yaml", wleName)] = fmt.Sprintf(staticConfig, si.Endpoint.Address)
+	}
+	// Write the update config map back to cluster
+	if _, err := c.client.CoreV1().ConfigMaps("istio-system").Update(context.TODO(), fileSDConfig,
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:847]: There is a great possibility to race when there are multi-istiod

**Comment by howardjohn** [comment]
We may want to consider an architecture where instead of writing to a configmap we have a sidecar to prometheus and just write to local disc. Not sure if its a good idea, just an idea

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 4 above)
Comment (ZhengzheYang) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:847]: Thanks for pointing this out! If we have multi-istiod, normally that's the case when we have multiple clusters? Then won't this line write to their own cluster? Or am I understanding it wrong? Thanks!

**Comment by ZhengzheYang** [comment]
> We may want to consider an architecture where instead of writing to a configmap we have a sidecar to prometheus and just write to local disc. Not sure if its a good idea, just an idea

@howardjohn I think this is a great idea, but @douglas-reid seemed to have tried this approach before we decided to go down this path.

**Comment by ZhengzheYang** [comment]
I would draft an RFC to kick off the discussion! Update soon.

**Comment by douglas-reid** [comment]
> @howardjohn I think this is a great idea, but @douglas-reid seemed to have tried this approach before we decided to go down this path.

To be clear, I have never tried adding adding a container that could run alongside prometheus to write out the file locally. Having an envoy sidecar has been used for varying purposes, including cert delivery.

In the design doc, we should talk about this alternative, and how it might work -- including the feasibility for bring-your-own scenarios.

**Comment by ZhengzheYang** [comment]
https://docs.google.com/document/d/1fzvyGa8y0d4aUixzPyPfh0dIPaThpRKwPF8ldboGnCA/edit?usp=sharing

@howardjohn @douglas-reid I draft an RFC where people can comment and discuss. Let me know if there's any missing parts and I would definitely add those. 

> > @howardjohn I think this is a great idea, but @douglas-reid seemed to have tried this approach before we decided to go down this path.
> 
> To be clear, I have never tried adding adding a container that could run alongside prometheus to write out the file locally. Having an envoy sidecar has been used for varying purposes, including cert delivery.
> 
> In the design doc, we should talk about this alternative, and how it might work -- including the feasibility for bring-your-own scenarios.

Oh, sorry for misunderstanding! I will definitely talk about those in the doc. Thanks!

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:847]: Multi instances can be deployed in one cluster, for HA

Comment (ZhengzheYang) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:847]: Understood! This is helpful for me. And this is just a proof of concept for now and we might go with different approaches here. Thanks for reviewing!

**Comment by rshriram** [comment]
why not try persistent volumes instead of config maps? We can tell people that they have to mount the same volume on both prom and istio

**Pull request closed without merging**